### PR TITLE
[#742] Add multiple listeners

### LINKED ISSUE #742: Listen on multiple sockets
Multiple IP addresses might be assigned to the single interface. Gateway should be able to listen on multiple addresses.

### PR DESCRIPTION
**Pull request by KirillovDenis**
close #742 
Signed-off-by: Denis Kirillov <denis@nspcc.ru>

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -95,7 +95,7 @@ $ neofs-s3-gw --healthcheck_timeout 15s --connect_timeout 1m --rebalance_interva
 ### Monitoring and metrics
```
Comment (masterSplinter01) [inline docs/configuration.md:95]: I think in `--help` and in line 49:
```
### Listening on address and TLS
You can make the gateway listen on specific address using the `--listen_address` option
```
we should mention that only one address can be set in this way. Or make it similar to `--peers`, but I'm not sure.

File: api/handler/put.go [hunk 2]
```diff
@@ -326,16 +326,16 @@ func getCopiesNumberOrDefault(metadata map[string]string, defaultCopiesNumber ui
 	return uint32(copiesNumber), nil
 }
 
-func (h handler) formEncryptionParams(header http.Header) (enc encryption.Params, err error) {
-	sseCustomerAlgorithm := header.Get(api.AmzServerSideEncryptionCustomerAlgorithm)
-	sseCustomerKey := header.Get(api.AmzServerSideEncryptionCustomerKey)
-	sseCustomerKeyMD5 := header.Get(api.AmzServerSideEncryptionCustomerKeyMD5)
+func (h handler) formEncryptionParams(r *http.Request) (enc encryption.Params, err error) {
```
Comment (masterSplinter01) [inline api/handler/put.go:329]: It seems like `h` is not used inside the function.

File: cmd/s3-gw/app.go [hunk 3]
```diff
@@ -505,6 +432,40 @@ func (a *App) startServices() {
...
+			zap.String("tls cert", serverInfo.TLS.CertFile), zap.String("tls key", serverInfo.TLS.KeyFile))
+		a.servers[i] = newServer(ctx, serverInfo, a.log)
+	}
+}
+
+func (a *App) updateServers() error {
+	serversInfo := fetchServers(a.cfg)
+
+	if len(serversInfo) != len(a.servers) {
+		return fmt.Errorf("invalid servers configuration: length mismatch: old '%d', new '%d", len(a.servers), len(serversInfo))
```
Comment (masterSplinter01) [inline cmd/s3-gw/app.go:451]: Maybe "amount mismatch" would be clearer than "length mismatch"?

**Review by alexvanin — COMMENTED** [review]
Rebase, please, and consider https://github.com/nspcc-dev/neofs-http-gw/pull/230#discussion_r1030575642 here as well.

**Comment by KirillovDenis** [comment]
> Rebase, please, and consider [nspcc-dev/neofs-http-gw#230 (comment)](https://github.com/nspcc-dev/neofs-http-gw/pull/230#discussion_r1030575642) here as well.

I'm not sure we really need this. We have to make `cfgTLSCertFile` and `cfgTLSKeyFile` also slice parameters and it doesn't allow us to set several listen addresses with different tls settings (e.g. we want the first address be without tls and the second with tls) using command line args. So I suggest doesn't change the flags here.

**Review by alexvanin — APPROVED** [review]

**Pull request merged**